dopamine/utils/lock.py

```diff
@@ -0,0 +1,58 @@
+# coding=utf-8
+# Copyright 2018 The Dopamine Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Creates a function decorator to protect execution with a lock."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import threading
+
+_LOCK_ATTR_NAME = '_lock'
+
+
+# TODO(aarg): Add a lock implementation that lets read operation occur at the
+# same time (i.e. less restrictive).
+class LockedClass(object):
```

Comment:
Something about the class and inheritance is uncomfortable for me.  I think it makes more sense for locks defined by composition, not inheritance.  Like that initialization of a `OutOfGraphReplayBuffer` also takes a `lock=None`.  Then `_LOCK_ATTR_NAME` would be a member of the implementing class, and the decorator would take it as an argument.

This makes locking more flexible (it doesn't just have to use threading.Lock()).  I somehow also want to say it's easier to test -- both in unit tests for this class and in unit tests for classes containing it?  But my ideas there are super hand-wavy.

In any event, what do you think?  Worth doing / not worth doing / worth waiting for psc@'s feedback?

Reply:
Done. I think it's a slight improvement. I'd love for the user to avoid to have to modify their code in order to apply the lock. It's still not the case yet.
Let's get psc@'s feedback.

Reply:
I also wanted to defer the lock object management to the lock module without having the user deal with it - which was the role of the parent class.
I am replacing that with a `lock_initialization` function. What do you think?